extensions/blocks/mailchimp/edit.js

```diff
@@ -190,6 +193,26 @@ class MailchimpSubscribeEdit extends Component {
 						onChange={ this.updateErrorText }
 					/>
 				</PanelBody>
+				<PanelBody title={ __( 'Mailchimp Groups', 'jetpack' ) }>
+					<MailchimpGroups
+						interests={ interests }
+						onChange={ ( id, checked ) => {
+							setAttributes( {
+								interests: [ ...new Set( [ ...interests, id ] ) ].filter( item =>
```

Comment:
I also think it's a little hard to grok; setting attributes and modifying them within the same block of code is one thing and another is this `[ ...new Set( [ ...interests, id ] ) ]` which takes a while to read.

It looks like you're building a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set), from an array and an item, then turning that again into an array?

Can some parts of this be done differently, e.g. using `add(id)`, returning iterable with `.entries()` or not using `Set` at all?

```js
const interests = new Set( [ ...interests, id ] ).entries();
// or
const interests = new Set(interests).add(id).entries();
// or
const interests = [...interests, id];

interests.filter( ... );
```

I'd also break the function out from JSX and refer to it with `onChange={ handleGroupChange }` or `onChange={() => this.handleGroupChange(id, checked) }` instead, but that's just personal preference. :-)

Thoughts?

Reply:
Totally get it, my stylish modern Javascript was just too much for you guys 😄. This was the leanest version I was able to come up with. I'm using `Set()` to insure there  are no duplicates, then adding (or removing)  the new value based on the checkbox value. I _think_ all of @simison's alternatives don't account for the value of the checkbox, and also would allow for dupes if they get in there somehow. Here's an alternative that avoids `Set()`. It's still kind of baroque, but might be a bit more readable? What do you think?

```
const keyed = interests.reduce( ( acc, item ) => ( { ...acc, [ item ]: true } ), {} ); // Convert array to a keyed object
keyed[ id ] = checked; // Set current value to true/false based on checked value;
const newInterests = Object.keys( keyed ).filter( item => keyed[ item ] ? item : null ); // Convert back to an array, omitting current id if the value is false.
```

Reply:
Yasss your code is too cool for us, haha! Okay so that makes sense now. :-)

One more option would be to use [`uniq`](https://lodash.com/docs/4.17.15#uniq) from Lodash.

```js
  const filteredInterests = uniq([ ...interests, id ]).filter( item =>
    item === id && ! checked ? false : item
  );
```

I think even if you just decide to use `Set()`, it would be helpful to have a comment above it that it's removing duplicates because it wasn't clear and thus confusing. Splitting the thing to a couple more lines gives also space to explain it step-by-step:

```js
const handleGroupChange = ( id, checked ) => {
  const newInterests = [ ...new Set( [ ...interests, id ] ) ];

  const filteredInterests = newInterests.filter( item =>
    item === id && ! checked ? false : item
  );

  setAttributes( { interests: filteredInterests } );
};
```


Up to you. :-)

Reply:
How's this look? https://github.com/Automattic/jetpack/pull/14050/commits/bf454be06c80c49895d83d435657fae95f77333f

Reply:
Looks great! 👍 thanks